app/src/main/java/com/vero/tasky/agenda/data/remote/repository/AgendaRepositoryImpl.kt

```diff
@@ -18,25 +24,65 @@ class AgendaRepositoryImpl(
     private val reminderDao: ReminderDao,
 ) : AgendaRepository {
 
-    override suspend fun getAgendaForDay(timezone: String, time: Long): Flow<Result<Agenda>> = flow {
+    override suspend fun getAgendaForDay(
+        timezone: String,
+        timestamp: Long
+    ) : Flow<List<AgendaItem>> = flow {
+
+        emit(getAgendaItemsFromDao(timestamp))
+
+        ApiCallHandler.safeApiCall {
+            val networkAgendaItems = api.getAgendaForDay(timezone, timestamp)
+            saveAgendaItems(networkAgendaItems)
+        }
+
+        emit(getAgendaItemsFromDao(timestamp))
     }
 
     override suspend fun syncAgenda(
         deletedEventsIds: List<String>,
         deletedTasksIds: List<String>,
         deletedReminderIds: List<String>
     ): Result<Unit> = ApiCallHandler.safeApiCall {
-            api.syncAgenda(
-                SyncAgendaRequest(
-                    deletedEventsIds = deletedEventsIds,
-                    deletedTasksIds = deletedTasksIds,
-                    deletedReminderIds = deletedReminderIds
-                )
+        api.syncAgenda(
+            SyncAgendaRequest(
+                deletedEventsIds = deletedEventsIds,
+                deletedTasksIds = deletedTasksIds,
+                deletedReminderIds = deletedReminderIds
             )
+        )
     }
 
-    override suspend fun getFullAgenda() =
-        ApiCallHandler.safeApiCall {
-            Agenda(emptyList())
+    override suspend fun getFullAgenda() = ApiCallHandler.safeApiCall {
+        val result = api.getFullAgenda()
+        saveAgendaItems(result)
+    }
+
+    private fun getAgendaItemsFromDao(timestamp: Long) : List<AgendaItem> {
+        val agendaItems = mutableListOf<AgendaItem>()
+        agendaItems.addAll(taskDao.loadTasksForDay(timestamp).map {
+            TaskMapper.toTask(it)
+        })
+        agendaItems.addAll(eventDao.loadEventsForDay(timestamp).map {
+            EventMapper.toEvent(it)
+        })
+        agendaItems.addAll(reminderDao.loadRemindersForDay(timestamp).map {
+            ReminderMapper.toReminder(it)
+        })
+        return agendaItems
+    }
+
+    private suspend fun saveAgendaItems(agendaDto: AgendaDto) {
+        supervisorScope {
+            agendaDto.events.map { eventDto ->
+                launch { eventDao.insertEvents(EventMapper.toEventEntity(eventDto)) }
+            }.forEach { job -> job.join() }
+            agendaDto.tasks.map { taskDto ->
+                launch { taskDao.insertTasks(TaskMapper.toTaskEntity(taskDto)) }
+            }.forEach { job -> job.join() }
+            agendaDto.reminders.map { reminderDto ->
+                launch { reminderDao.insertReminders(ReminderMapper.toReminderEntity(reminderDto)) }
+            }.forEach { job -> job.join() }
```

Comment:
Joining these jobs should happen at the end of the function for all and not after every step. In your implementation all events are inserted in parallel, but it will resume with inserting tasks only when all events have been inserted. Instead just save the references to the jobs and then join them all at the end